File: TOP150Review/Best Time to Buy and Sell Stock.py

class Solution:

    # 迴圈保持一個指標說明到此前最便宜買入價格 , 在每一格計算最大profit 
    def maxProfit(self, prices: List[int]) -> int:
        
        best = 0 
        min_buy = prices[0] 

        for i in range(1 , len(prices)):  

            best = max(best  , prices[i] - min_buy)  
            min_buy = min(min_buy , prices[i]) 
        
        return best 


        

File: RUSH150/Best Time to Buy and Sell Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {

    const prefix = new Array(prices.length) ;
    prefix[0] = prices[0] 
    for (let i = 1 ; i < prices.length ; i++){
        prefix[i] = Math.min( prefix[i-1] , prices[i] )  
    }
    
    let solution = 0 ; 
    for (let i = 0 ; i < prices.length ; i++){
        
        solution = Math.max( solution , prices[i] - prefix[i] )
    }
    return solution ;    
};


var maxProfit = function(prices) {

    let minPrices = Number.MAX_VALUE ; 
    let solution = 0 ; 

    for (let i = 0 ; i < prices.length ; i++){
        minPrices = Math.min(minPrices , prices[i]) ; 
        solution = Math.max(  solution , prices[i] - minPrices )

    }
    return solution ;    
};